dds/src/dcps/infrastructure/instance.rs

```diff
@@ -1,36 +1,123 @@
-use core::ops::Index;
+use crate::transport::types::Guid;
+use core::{
+    borrow::Borrow,
+    ops::{Deref, Index},
+};
 use dust_dds_derive::TypeSupport;
 
-/// Type for the instance handle representing an Entity
+/// Special constant value representing a 'nil' [`InstanceHandle`].
+pub const HANDLE_NIL: InstanceHandle = InstanceHandle([0; 16]);
+
+/// Type for the instance handle representing an Entity.
+#[repr(transparent)]
 #[derive(Clone, Copy, PartialEq, Eq, Debug, Hash, PartialOrd, Ord, TypeSupport)]
 pub struct InstanceHandle([u8; 16]);
 
 impl InstanceHandle {
-    /// InstanceHandle constructor
+    /// Constructs a new `InstanceHandle`.
+    #[inline]
     pub const fn new(bytes: [u8; 16]) -> Self {
-        InstanceHandle(bytes)
+        Self(bytes)
+    }
+}
+
+impl From<[u8; 16]> for InstanceHandle {
+    #[inline]
+    fn from(value: [u8; 16]) -> Self {
+        Self(value)
+    }
+}
+
+impl From<InstanceHandle> for [u8; 16] {
+    #[inline]
+    fn from(value: InstanceHandle) -> Self {
+        value.0
+    }
+}
+
+impl From<Guid> for InstanceHandle {
+    #[inline]
+    fn from(value: Guid) -> Self {
+        Self(value.into())
+    }
+}
+
+impl From<InstanceHandle> for Guid {
+    #[inline]
+    fn from(value: InstanceHandle) -> Self {
+        value.0.into()
     }
 }
 
 impl From<crate::rtps::behavior_types::InstanceHandle> for InstanceHandle {
+    #[inline]
     fn from(value: crate::rtps::behavior_types::InstanceHandle) -> Self {
         Self(value.0)
     }
 }
 
 impl From<InstanceHandle> for crate::rtps::behavior_types::InstanceHandle {
+    #[inline]
     fn from(value: InstanceHandle) -> Self {
-        crate::rtps::behavior_types::InstanceHandle(value.0)
+        Self(value.0)
+    }
+}
+
+impl PartialEq<[u8; 16]> for InstanceHandle {
+    #[inline]
+    fn eq(&self, other: &[u8; 16]) -> bool {
+        self.0.eq(other)
+    }
+}
+
+impl PartialEq<InstanceHandle> for [u8; 16] {
+    #[inline]
+    fn eq(&self, other: &InstanceHandle) -> bool {
+        self.eq(&other.0)
+    }
+}
+
+impl PartialEq<[u8]> for InstanceHandle {
+    #[inline]
+    fn eq(&self, other: &[u8]) -> bool {
+        other
+            .try_into()
+            .is_ok_and(|other: &[u8; 16]| self.eq(other))
+    }
+}
+
+impl PartialEq<InstanceHandle> for [u8] {
+    #[inline]
+    fn eq(&self, other: &InstanceHandle) -> bool {
+        other.eq(self)
+    }
+}
+
+impl Deref for InstanceHandle {

```

Comment:
Deref should exclusively be used for types that represent smart pointers so this implementation should be remove. We have tried this approach before and it doesn't end up very well.